### PR TITLE
feat(resume): rewrite /resume to list all history sessions automatically

### PR DESCRIPTION
**Pull request by randomuser2026x**
## Problem

`/resume` could only restore sessions explicitly named via `/save`. Unnamed sessions — the vast majority in gateway mode (Feishu, Telegram, Discord, etc.) — were invisible and lost forever after context compression or session reset.

## Solution

Rewrite `/resume` to automatically discover and list **all** past sessions using a new **SessionTree** architecture.

### New behavior

```
User:  /resume
Bot:   Select a session to resume:
       [1] 2h ago — 帮我分析一下这个报错日志
       [2] 3d ago — Rewrite the authentication module
       [3] 5d ago — [named: refactor-api] API refactoring
       
User:  /resume 1
Bot:   ✅ Resumed session (15 messages restored)
```

- **`/resume`** — numbered list of all previous sessions with timestamps and previews
- **`/resume last`** — instantly resume the most recent session
- **`/resume <name>`** — resume by name or list number (unchanged)
- Compression chains deduplicated — only the latest leaf is shown

### Architecture: SessionTree

Before: N+1 database queries per resume (one per ancestor + one per session to find compression leaf).

After: **1 SQL query** loads all sessions → in-memory tree → Python traversal.

```
SessionNode (dataclass)
  ├── parent_id, children[]
  ├── is_compressed, end_reason
  └── last_active, message_count, first_message

SessionTree
  ├── build_session_tree()          # 1 SQL query → full tree
  ├── get_resume_candidates()       # unified candidate logic
  ├── find_compression_leaf()       # follow compression chain only
  └── get_same_conversation_ids()   # proper exclusion scope
```

## Commits

**Commit 1: feat(resume): rewrite /resume to list all history sessions** (6 files, +1,305/-60)
- `hermes_state.py` — SessionNode, SessionTree, build_session_tree(), get_resume_candidates(), find_compression_leaf(), get_same_conversation_ids()
- `gateway/run.py` — /resume command rewrite using SessionTree, `/resume last` shortcut, per-platform candidate maps, configurable resume_min_messages (-136 net lines)
- `gateway/session.py` — crash recovery (auto-follow compression leaf on restart), on_session_split callback for atomic sessions.json sync, threading.Lock deadlock fix
- `SESSION_TREE_PLAN.md` — architecture design document

**Commit 2: fix(compression): complete compression coverage** (1 file, +172/-15)
- `run_agent.py` — compression now triggers on ALL API response paths (was only tool_calls branch), preflight uses real API token count instead of chars/4

## Testing

- 9 new unit tests for SessionTree edge cases
- Verified against production state.db (15+ previously misidentified nodes corrected)
- Gateway running with these changes, no regressions

### CODE REVIEW

File: gateway/run.py [hunk 1]
```diff
@@ -6843,58 +6908,184 @@ async def _handle_resume_command(self, event: MessageEvent) -> str:
...
+
+        current_entry = self.session_store.get_or_create_session(source)
+        current_sid = current_entry.session_id
+
+        # Check if we're dealing with a placeholder entry (empty session_id)
+        # This happens after gateway restart when the session_key wasn't in _entries
+        _is_placeholder_entry = _is_placeholder_entry or not current_entry.session_id
+
+        # Build session tree once for all operations
+        tree = self._session_db.build_session_tree(source=user_source)
```
Comment (teknium1) [inline gateway/run.py:6926]: This platform-only tree can include sessions from other users or rooms on the same adapter, and the later switch path has no ownership check. Current main requires per-row visibility filtering and `_resume_target_allowed` before switching (`gateway/slash_commands.py:793-958`) to prevent cross-origin transcript enumeration and resume; preserve that guard when porting this feature.

File: run_agent.py [hunk 2]
```diff
@@ -8907,13 +8914,24 @@ def run_conversation(
...
-            )
+            # Use last_prompt_tokens (real API value) if available, fall back
+            # to estimate only if stale (0) after disconnects or new sessions.
+            # The rough estimate underestimates by ~10-15% because it misses
+            # system prompt (~4K) and tool schemas (~14K), causing premature
+            # compression loops.  (#2153)
+            _cc = self.context_compressor
+            if _cc.last_prompt_tokens > 0:
+                _preflight_tokens = (
+                    _cc.last_prompt_tokens + _cc.last_completion_tokens
```
Comment (teknium1) [inline run_agent.py:8925]: Do not add completion/reasoning tokens to request-context pressure. They do not occupy the next prompt's context window and can be large for reasoning models; current main deliberately uses `last_prompt_tokens` only at `agent/conversation_loop.py:4748-4754` to avoid premature compression.

**Review by teknium1 — COMMENTED** [review]
Thanks for tackling gateway recovery and session discoverability. The default `/resume` premise is still present on current main (`gateway/slash_commands.py:3544-3560`), although `/sessions full` now provides a scoped unnamed-session browser (`gateway/slash_commands.py:3728-3747`).

### Problems
- The PR lists candidates using only `source=user_source` (`gateway/run.py:6926` on the PR head), then switches at `gateway/run.py:7090` without proving caller ownership. Current main added the required fail-closed list and target guards in `gateway/slash_commands.py:793-958` via `c4f278c02`; salvaging this must retain them.
- The PR treats completion/reasoning tokens as next-request context pressure (`run_agent.py:8923-8926` on the PR head). Current main intentionally uses prompt tokens only because completion/reasoning tokens do not consume the next request context (`agent/conversation_loop.py:4748-4754`).
- The handler moved to `gateway/slash_commands.py:3517`, and current `list_sessions_rich` already projects compressed lineages to their tip (`tests/test_hermes_state.py:3839-3866`), so this is not a clean cherry-pick.

### Suggested changes
- Rebuild the feature in `GatewaySlashCommandsMixin` using the current listing helpers and preserve ownership checks for both enumeration and switching.
- Add cross-user/cross-chat visibility tests and keep compression work separate.

Automated hermes-sweeper review.
